src/cascadia/Remoting/Monarch.h

```diff
@@ -69,14 +70,19 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
     private:
         uint64_t _ourPID;
 
-        uint64_t _nextPeasantID{ 1 };
-        uint64_t _thisPeasantID{ 0 };
+        std::atomic<uint64_t> _nextPeasantID{ 1 };
+        uint64_t _ourPeasantId{ 0 };
+
+        // When we're quitting we do not care as much about handling some events that we know will be triggered
+        std::atomic<bool> _quitting{ false };
 
         winrt::com_ptr<IVirtualDesktopManager> _desktopManager{ nullptr };
 
         std::unordered_map<uint64_t, winrt::Microsoft::Terminal::Remoting::IPeasant> _peasants;
+        std::shared_mutex _peasantsMutex{};

```

Comment:
I did originally consider it, but disregarded because the first iteration of this had a single `std::recursive_mutex _mutex` that was used for locking both `_peasants` and `_mruPeasants`. Of course the current iteration does not have that, and I was able to separate out (and downgrade to) a `std::shared_mutex`. 

Once we are confident that this locking strategy is what we want I can take another look at using `til::shared_mutex`, although I don't look forward to updating all of the remoting tests that rely on friend access.

Reply:
If possibly I would like this refactor to be deferred for now. I took a shot at doing so, but it proved to be a bit more difficult than I hoped.